### DISCUSSION TITLE
Tajimas D

### DISCUSSION THREAD

**Discussion by Didoule** [cogent3/cogent3#1224] [Ideas]
Hello - firstly I'd just like to say I am really enjoying using this set of tools for analysing my sequence data. Thanks very much for putting this together. 

I would like to calculate Tajima's D for my set of sequences and wondered if this was possible using Cogent3? I couldn't find this in the available statistical models. If possible I would like to obtain all pairwise comparisons.
I'm also interested in calculating the dN/dS ratio per codon along a sequence. There is a program written in R which does this (called OmegaMap) but it struggles with large datasets. I wondered if this was something that could be developed for Cogent3 (might be quite computationally challenging though) so no problem if not possible. This kind of analysis would allow codons under immune selection (or other) to be detected.  

Many thanks
Odile

**Comment by GavinHuttley** [comment]
> Is Tajima's D implemented?

No, but you can build your own using the `variable_positions()` method on alignments.

> Is it possible to calculate the dN/dS ratio per codon along a sequence

You would need to do codon modelling, i.e. constructing a likelihood function. It's quite trivial to do a sliding window type analysis (the `Alignment` object has a `sliding_windows()` method, you would need to ensure `step=3` to stay in frame). As you say, that would be quite compute intensive. It can be run in parallel, but whether that's fruitful depends on how many alignments you're dealing with.

An alternate would be to employ a [Phylo-HMM](https://cogent3.org/doc/examples/hmm_par_heterogeneity.html). It is possible to apply that approach to an arbitrary parameter, including `omega` from a codon model. Doing so does not explicitly address estimation of codon position specific parameter `omega` value but could rather deliver a posterior probability of the position belong to one class of sites over another, etc... This approach is non-trivial and requires an appreciation for the statistical challenges of fitting such a model and how those will likely play out with the data you are analysing. (A fuller answer is probably an entire research paper!)
